modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsManagerImpl.java

```diff
@@ -278,4 +739,349 @@ private ObjectStatisticsImpl aggregateLocalStatistics(
 
         return tblStats;
     }
+
+    /**
+     * Receive and handle statistics propagation message with partitions statistics.
+     *
+     * @param nodeId Sender node id.
+     * @param msg Statistics propagation message with partitions statistics to handle.
+     * @throws IgniteCheckedException In case of errors.
+     */
+    private void receivePartitionsStatistics(UUID nodeId, StatsPropagationMessage msg) throws IgniteCheckedException {
+        UUID locNode = ctx.discovery().localNode().id();
+        for (StatsObjectData partData : msg.data()) {
+            StatsKey key = new StatsKey(partData.key().schema(), partData.key().obj());
+
+            assert partData.type() == StatsType.PARTITION : "Got non partition level statistics by " + key
+                    + " without request";
+
+            if (log.isTraceEnabled())
+                log.trace(String.format("Received partition statistics %s.%s:%d from node %s", key.schema(),
+                        key.obj(), partData.partId(), nodeId));
+
+            GridH2Table tbl = schemaMgr.dataTable(key.schema(), key.obj());
+            if (tbl == null) {
+                if (log.isInfoEnabled())
+                    log.info(String.format("Ignoring outdated partition statistics %s.%s:%d from node %s",
+                            key.schema(), key.obj(), partData.partId(), nodeId));
+
+                continue;
+            }
+            GridDhtPartitionState partState = tbl.cacheContext().topology().partitionState(locNode, partData.partId());
+            if (partState != OWNING) {
+                if (log.isInfoEnabled())
+                    log.info(String.format("Ignoring non local partition statistics %s.%s:%d from node %s",
+                            key.schema(), key.obj(), partData.partId(), nodeId));
+
+                continue;
+            }
+
+            ObjectPartitionStatisticsImpl opStat = StatisticsUtils.toObjectPartitionStatistics(ctx, partData);
+
+            statsRepos.saveLocalPartitionStatistics(key, opStat);
+        }
+    }
+
+    /**
+     * Receive and handle statistics propagation message as response for collection request.
+     *
+     * @param nodeId Sender node id.
+     * @param msg Statistics propagation message with partitions statistics to handle.
+     * @throws IgniteCheckedException In case of errors.
+     */
+    private void receiveLocalStatistics(UUID nodeId, StatsCollectionResponse msg) {
```

Comment:
What does mean 'local' in method name?

Reply:
"Level" of statistics. There are partition, local and global statistics. Yes, strictly it can be not exactly local statistics, if there are some rescheduling during whole process, but in general receiveLocalStatistics method should receive collected "local" statistics and try to finalyze statistics collection process.